lms/djangoapps/courseware/field_overrides.py

```diff
@@ -76,13 +77,29 @@ def wrap(cls, user, wrapped):
                  settings.FIELD_OVERRIDE_PROVIDERS))
 
         if cls.provider_classes:
-            return cls(user, wrapped)
+            # TODO: we might not actually want to return here.  Might be better
+            # to check for instance.providers after the instance is built. This
+            # would allow for the case where we have registered providers but
+            # none are enabled for the provided course
+            return cls(user, wrapped, course)
 
         return wrapped
 
-    def __init__(self, user, fallback):
+    def __init__(self, user, fallback, course):
         self.fallback = fallback
-        self.providers = tuple((cls(user) for cls in self.provider_classes))
+        # set self.providers to be a filtered list of enabled providers based
+        # on the course passed in.  Cache this result per request to avoid
+        # needing to call the provider filter api hundreds of times
+        request_cache = RequestCache.get_request_cache()
+        enabled_providers = request_cache.data.get(
+            ENABLED_OVERRIDE_PROVIDERS_KEY, NOTSET
+        )
+        if enabled_providers == NOTSET:
+            enabled_providers = tuple(
+                (cls(user) for cls in self.provider_classes if cls.enabled_for(course))
+            )
+            request_cache.data[ENABLED_OVERRIDE_PROVIDERS_KEY] = enabled_providers
```

Comment:
This assumes that any given request will only be instantiating XBlocks for a single course. That's true now, I think, but I'm not sure it'll always be true. It might be better to use the `CourseKey` (`course.id`) as part of the key in the request cache.